### PR TITLE
Add EdDSA support to EVP_PKEY_sign/verify

### PR DESCRIPTION
**Pull request by jvdsn**
As mentioned in https://github.com/openssl/openssl/pull/22112, it would be nice if EdDSA was accessible using the EVP_PKEY_sign/verify APIs. Right now, EdDSA is actually quite counter-intuitive, because the DigestSign/Verify APIs need to be used, but the md must be set to NULL, and only the one-shot API works.

Let me know if I missed something, these changes seemed pretty straightforward but I'm not very familiar with the many abstraction layers in OpenSSL.

I updated the tests and they're passing locally. I confirmed they were actually being executed by printing some debug messages. If you decide to accept this, I will work on updating the docs.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: test/recipes/30-test_evp_data/evppkey_ecx.txt
Comment (slontis) [inline test/recipes/30-test_evp_data/evppkey_ecx.txt:1178]: I think you want >=3.3.0 ?

Comment (jvdsn) [reply test/recipes/30-test_evp_data/evppkey_ecx.txt:1178]: Yeah, you're right, and a 3.3.0 guard should also be added to the other new tests. I'll work on that.

**Comment by t8m** [comment]
Yeah, this is something we definitely want to add. It would be also interesting if we could add a compat code in the EVP layer that would for the algorithms that do not use EVP_MD but do not implement sign()/verify() tried calling digestsign() digestverify() instead. Otherwise the application cannot really use EVP_PKEY_sign/verify with EdDSA unless it knows that it uses 3.3 fips provider (which probably won't be FIPS validated).

**Comment by jvdsn** [comment]
@t8m is PKEY_sign/verify always supposed to behave the same as DigestSign/Verify with no MD?

Also, regarding the FIPS provider, isn't that a more general issue? Let's say 3.1.x is validated, so that .so file is dropped in for OpenSSL 3.1.y. How does the application know it cannot use any 3.1.y features?

**Comment by levitte** [comment]
> Yeah, this is something we definitely want to add. It would be also interesting if we could add a compat code in the EVP layer that would for the algorithms that do not use EVP_MD but do not implement sign()/verify() tried calling digestsign() digestverify() instead. Otherwise the application cannot really use EVP_PKEY_sign/verify with EdDSA unless it knows that it uses 3.3 fips provider (which probably won't be FIPS validated).

This could be tricky, 'cause in libcrypto, there are fallbacks to EVP_PKEY_sign/verify (they were originally designed to give support for "primitive" operations).

File: doc/man7/EVP_SIGNATURE-ED25519.pod
Comment (t8m) [inline doc/man7/EVP_SIGNATURE-ED25519.pod:232]: Please add a HISTORY section that would mention that this is new functionality in 3.3.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
@jvdsn the backwards compat handling is something to think about in future, it should be in a separate PR.

File: doc/man7/EVP_SIGNATURE-ED25519.pod
Comment (t8m) [reply doc/man7/EVP_SIGNATURE-ED25519.pod:232]: Also please add a CHANGES.md entry.

**Comment by t8m** [comment]
Please drop the merge commit. If it is necessary because of conflicts, please rebase the pull request against fresh master branch.

**Comment by jvdsn** [comment]
> Please drop the merge commit. If it is necessary because of conflicts, please rebase the pull request against fresh master branch.

Yeah, there was a conflict. I rebased it now.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Ping @openssl/committers for second review

**Comment by bernd-edlinger** [comment]
I think creates an overload to the `EVP_PKEY_sign` and `EVP_PKEY_verify` function in  quite an error prone way:
These function was always expecting a MessageDigest in the `tbs` parameter, but now it is no longer
clear if the `tbs` parameter is a MessageDigest or the message itsef.
So If you pass the wrong data you will get an ECDSA of only the first few bytes of the Message
are used to create the signature.
If you accidentally pass the MessageDigest to an EDDSA key you will get the EDDSA of the
MessageDigest which is not what you want either.
The function itself has no way to check that the correct value is given.

**Comment by davidben** [comment]
+1 to @bernd-edlinger. This PR will make OpenSSL's API extremely dangerous and break having an algorithm-generic signing API. This was why we intentionally didn't wire it up to EVP_PKEY_sign at the time. I'd suggested EVP_PKEY_sign_message but OpenSSL preferred to tie it to EVP_MD_CTX. It's unfortunate that the names read backwards from what they actually are, but if OpenSSL values backwards compatibility, it can't change the meanings of those names now.

Mixing up signing and prehashed signing APIs is not safe. If you pass an unhashed input to a prehashed API, if it happens to be short enough to fit (or e.g. ECDSA which truncates), the result is not a secure signature scheme. The hashes are used for more than just shrinking the input.

**Comment by t8m** [comment]
Using the EVP_MD_CTX and EVP_DigestSign API for algorithms where the digest is inherent part of the algorithm and cannot be really a separate step is very confusing and bad as well. The EVP_DigestSign API should have never been overloaded with the EdDSA support.

We could add EVP_PKEY_sign_message instead as a separate call and block it for RSA, ECDSA and DSA algorithms.

**Comment by davidben** [comment]
@t8m Please see https://github.com/openssl/openssl/pull/3409 for the previous discussion. The reason you all choose to use `EVP_DigestSign` was for consistency with your existing APIs.

You could add a new API if you no longer like your existing APIs, but disabling the new API with RSA, ECDSA, and DSA would defeat the purpose. The whole point of having common APIs between algorithms is so that the same function will perform equivalent operations across those algorithms. Callers who want to support ECDSA and Ed25519 in the same context (e.g. OpenSSL's own X.509 and TLS implementations) depend on this.

Now, for a signature scheme, there are three flavors of the operation you might care about supporting:

1. Signing an input. This is the actual signing operation that _every_ signature scheme will provide, or it's not a signature scheme. In [RSASSA-PSS-SHA256](https://www.rfc-editor.org/rfc/rfc8017.html), this signing operation is to hash then encode with EMSA-PSS, then perform the RSASP1 transform. OpenSSL likes to decompose these three steps in its APIs, but it should be thought of as one combined operation, because the security properties only hold when used together. In Ed25519, this is the atomic [Ed25519](https://www.rfc-editor.org/rfc/rfc8032.html) operation you're familiar with.

2. Signing an input, but providing the message in a streaming matter with an Init/Update/Final-style API. This is semantically equivalent to the first operation, as it should always produce the same results as concatenating and then signing, but not all signature schemes admit efficient implementations of this operation. Ed25519, notably does not.

3. Signing a pre-hashed input. This is a subtle operation as it decomposes the true signature scheme into two steps, and defers one step to the caller. The caller is assumed to have run it through the correct hash function because, otherwise, the result is not a secure signature scheme. Not all signature schemes support this operation. Ed25519, again, does not.

These three operations, particularly the third one, are different. It is important that OpenSSL's APIs are clear on which operation is being performed and never mix them. When and _only when_ the operation does not work for the signature scheme should the operation fail. Currently, OpenSSL's APIs for these are:

1. `EVP_DigestSignInit` + `EVP_DigestSign`
2. `EVP_DigestSignInit` + `EVP_DigestUpdate` + `EVP_DigestSignFinal`
3. `EVP_PKEY_sign_init` + `EVP_PKEY_sign`

It is unfortunate the one that actually signs a digest is the only one that doesn't say "digest" in the name, but so it goes.

Making the API for (3) suddenly pick up (1)'s semantics based on the key type, as this PR does, would be dangerous. You should be able to look at calling code and, without knowing the type of the key, be able to say which high-level operation is being performed. Otherwise the common APIs become a risk, not a benefit.

Making a new API for (1) that doesn't work for all algorithms, as you suggest, makes no sense, because every signature scheme supports (1).

**Comment by jvdsn** [comment]
It doesn't look like this PR will be accepted, so I'm closing it. EVP_DigestSign and EVP_DigestVerify will have to be exposed to the FIPS provider for the EdDSA self-tests.

**Comment by t8m** [comment]
> It doesn't look like this PR will be accepted, so I'm closing it. EVP_DigestSign and EVP_DigestVerify will have to be exposed to the FIPS provider for the EdDSA self-tests.

It has not been decided yet.

**Comment by t8m** [comment]
**OTC: How do we resolve the issue with misusing the EVP_DigestSign API for signature algorithms which do not really provide separate Digest then Sign operations?**

Using EVP_MD_CTX for operations that do not really operate this way is a complete mess. Given @davidben writeup above: https://github.com/openssl/openssl/pull/23240#issuecomment-1917532043 we probably should not overload the existing EVP_PKEY_sign() API which we might want to potentially deprecate or use only for the prehash versions of these algorithms.

**Comment by t8m** [comment]
One option would be to add:

- EVP_PKEY_sign_hash as an alias to the existing EVP_PKEY_sign.
- EVP_PKEY_sign_data (with EVP_PKEY_sign_data_update/final for streaming) as a new operation. This could even have a backwards compatibility backend implemented internally via the existing digestsign calls for providers that would not support the new operation.

**Comment by davidben** [comment]
I guess then the question is what problem we're trying to solve. You already have an API for `EVP_PKEY_sign_data` and streaming `EVP_PKEY_sign_data`, as well as other APIs designed around that API, like `X509_sign_ctx`. That was one of the considerations discussed in https://github.com/openssl/openssl/pull/3409.

I'd originally implemented `EVP_PKEY_sign_message` for BoringSSL, which was exactly the `EVP_PKEY_sign_data` idea. OpenSSL, however, preferred to put it in `EVP_MD_CTX`, which avoided the need to redo all the other APIs. It's a little goofy, but it does work. I then switched BoringSSL to align with OpenSSL's decision. See that PR for some of the history.

It sounds like you all now prefer to switch back to BoringSSL's original version. I'm curious why.

**Comment by t8m** [comment]
Because the DigestSign API is confusing and misleading and does not align with the provider APIs. Main reason for that confusion is that the EVP_MD_CTX is the primary context object for an operation that does not operate as hash.

Doing it that way was a gross hack and a serious mistake.

**Comment by t8m** [comment]
**OTC: Keeping this PR on hold until the design PR https://github.com/openssl/openssl/pull/22672 is finalized.**

**Comment by levitte** [comment]
> These functions were always expecting a MessageDigest in the `tbs` parameter, but now it is no longer clear if the `tbs` parameter is a MessageDigest or the message itsef.

Actually, this assumption about *tbs* is false, or rather, conditional to other functions being called, such as `EVP_PKEY_CTX_set_signature_md()` (and for RSA specifically, also requires setting a padding other than `RSA_NO_PADDING`).  In other words, the exact semantics depend on the provider implementation.  This is how it's documented.

> So if you pass the wrong data to an ECDSA key, you will get an ECDSA but only the first few bytes of the Message are used to create the signature. If you accidentally pass the MessageDigest to an EDDSA key, you will get the EDDSA of the MessageDigest which is not what you want either. The function itself has no way to check that the correct value is given.

Yes.  This is already the current situation.  Our API can only hope that the user sets the correct parameters for their use case, and provide a correct *tbs* (the only possible check is that *tbslen* is the right length for whatever input is expected).

**Comment by levitte** [comment]
> @t8m Please see #3409 for the previous discussion. The reason you all choose to use `EVP_DigestSign` was for consistency with your existing APIs.
> 
> You could add a new API if you no longer like your existing APIs, but disabling the new API with RSA, ECDSA, and DSA would defeat the purpose.

I don't recall suggesting removing anything.  `EVP_DigestSign` and friends are useful for any application that wants to composite a complete algorithm themselves.

However, as soon as there's a composite signature algorithm (like sha1WithRSAEncryption) being specified, everyone is currently forced to make a split-and-join operation (for example, to split it into sha1 and RSA) just to rejoin them in the EVP_DigestSignInit call, when it would be much more straighforward to fetch sha1WithRSAEncryption and use it, leaving it to the provider to do the right thing (which shouldn't be very hard, BTW, considering they more or less do that already).

**Comment by davidben** [comment]
> Actually, this assumption about tbs is false, or rather, conditional to other functions being called, such as EVP_PKEY_CTX_set_signature_md() (and for RSA specifically, also requires setting a padding other than RSA_NO_PADDING). In other words, the exact semantics depend on the provider implementation. This is how it's documented.

No, this confusion is *precisely* why mixing pre-hashed inputs and signing inputs in the same APIs is so dangerous. Hashing in a signature scheme is *not* merely for shrinking the message. It is a critical part of the security of the signature scheme.

If you pass a signing input to a signing function that expects a message digest, the result is not an alternate method for signing something. The result is a security vulnerability. E.g., for RSA, without hashing, your scheme is vulnerable to existential forgery.

This is why, in BoringSSL, we made sure to [carefully document](https://boringssl.googlesource.com/boringssl/+/940475da0c607eddc7c3e336284613f94632d99e%5E%21/) which APIs assume a pre-hash and which do not. SignMessage() and SignDigest() may have the same type signatures, but they are not the same function. To be a secure provider system, OpenSSL needs to carefully document, in both its application and provider interface, which one is which and never mix them up.

**Comment by levitte** [comment]
@davidben, I hear ya, and I share the security concerns.  But that doesn't change the current design of `EVP_PKEY_sign()`.

Re SignMessage() and SignDigest(), I assume those are names you came up with now?  After having discussed the ideas and names *at length*, I gotta say that I like those (apart from the camel casing) more than "primitive" and "data" (see #22672), so might borrow them...

**Comment by levitte** [comment]
> > But that doesn't change the current design of EVP_PKEY_sign().
> 
> Hmm. Is this an OpenSSL 3.x regression? At least in the OpenSSL 1.1.x days, `EVP_PKEY_sign` was always SignDigest and never SignMessage.

`EVP_PKEY_sign` itself makes no assumption of that sort, and never did.  The backend algorithm implementation, though, may care, given all sorts of conditions (including diverse defaults).

What can be said, though, is that the backends so far have implemented a mix of what you call SignDigest and what I would call "pure primitive" (in the RSA case, pretty close to RSASP1 for signatures), all depending on conditions (padding flags, whether a hash algo is specified, ...).  So no, so far, we've had no SignMessage implementation...  and given the outcry here, I can understand why that is.

> > Re SignMessage() and SignDigest(), I assume those are names you came up with now? After having discussed the ideas and names at length, I gotta say that I like those (apart from the camel casing) more than "primitive" and "data" (see #22672), so might borrow them...
> 
> Just ad-hoc names to try to describe the abstract operations. :-) Really I would have just called them Sign and SignDigest, but that get's confusing with `EVP_PKEY_sign`'s naming.

I like'em!

**Comment by t8m** [comment]
OTC: This is likely to be replaced by #24975

@jvdsn Can we close this? Please review the PR #24975 to see if it is OK for you.

**Comment by jvdsn** [comment]
@t8m yes, I believe this can be closed then. Thanks for the reviews

**Pull request closed without merging**
